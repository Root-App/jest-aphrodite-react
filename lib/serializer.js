'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
var cssUtil_1 = require('./cssUtil');
var reactUtils_1 = require('./reactUtils');
var replaceClassNames_1 = require('./replaceClassNames');
function createSerializer(getStyleSheetTestUtils, _a) {
  var _b = _a === void 0 ? {} : _a,
    _c = _b.removeVendorPrefixes,
    removeVendorPrefixes = _c === void 0 ? false : _c,
    classNameReplacer = _b.classNameReplacer;
  function test(val) {
    return (
      val &&
      !val.withStyles &&
      (val.$$typeof === Symbol.for('react.test.json') ||
        (val instanceof HTMLElement && !isBeingSerialized(val)))
    );
  }
  function print(val, printer) {
    var nodes = reactUtils_1.getNodes(val);
    nodes.forEach(function(node) {
      return (node.withStyles = true);
    });
    var selectors = cssUtil_1.getSelectors(nodes);
    var styles = cssUtil_1.getStyles(
      selectors,
      getStyleSheetTestUtils,
      removeVendorPrefixes,
    );
    var printedVal = printer(val);
    if (styles) {
      return replaceClassNames_1.replaceClassNames(
        selectors,
        styles,
        printedVal,
        classNameReplacer,
      );
    } else {
      return printedVal;
    }
  }
  return {
    test: test,
    print: print,
  };
}
exports.createSerializer = createSerializer;
function isBeingSerialized(node) {
  var currentNode = node;
  while (currentNode) {
    if (currentNode.withStyles) {
      return true;
    }
    currentNode = currentNode.parentNode;
  }
  return false;
}
// doing this to make it easier for users to mock things
// like switching between development mode and whatnot.
var getAphroditeStyleSheetTestUtils = function(useImportant) {
  return function() {
    return require('aphrodite' + (useImportant ? '' : '/no-important'))
      .StyleSheetTestUtils;
  };
};
exports.serializer = function(useImportant) {
  return createSerializer(getAphroditeStyleSheetTestUtils(useImportant));
};
