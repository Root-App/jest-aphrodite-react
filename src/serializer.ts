import { StyleSheetTestUtils } from 'aphrodite';
import * as css from 'css';
import { ReactTestRendererJSON } from 'react-test-renderer';
import { ClassNameReplacer, replaceClassNames } from './replaceClassNames';

export function createSerializer(
  getStyleSheetTestUtils: () => typeof StyleSheetTestUtils,
  classNameReplacer?: ClassNameReplacer,
): jest.SnapshotSerializerPlugin {
  function test(val: any) {
    return (
      val && !val.withStyles && val.$$typeof === Symbol.for('react.test.json')
    );
  }

  function print(val: ReactTestRendererJSON, printer: (val: any) => string) {
    const nodes = getNodes(val);
    nodes.forEach((node: any) => (node.withStyles = true));

    const selectors = getSelectors(nodes);
    const styles = getStyles(selectors);
    const printedVal = printer(val);
    if (styles) {
      return replaceClassNames(
        selectors,
        styles,
        printedVal,
        classNameReplacer,
      );
    } else {
      return printedVal;
    }
  }

  function getNodes(
    node: ReactTestRendererJSON,
    nodes: ReactTestRendererJSON[] = [],
  ) {
    if (node.children) {
      const children: ReactTestRendererJSON[] =
        typeof node.children === 'function'
          ? (node as any).children()
          : node.children;

      children.forEach(child => getNodes(child, nodes));
    }

    if (typeof node === 'object') {
      nodes.push(node);
    }

    return nodes;
  }

  function getSelectors(nodes: ReactTestRendererJSON[]) {
    return nodes.reduce<string[]>((selectors, node) => {
      const props =
        typeof node.props === 'function' ? (node as any).props() : node.props;
      return getSelectorsFromProps(selectors, props);
    }, []);
  }

  function getSelectorsFromProps(selectors: string[], props: any) {
    const className: string = props.className || props.class;
    if (className) {
      selectors = [
        ...selectors,
        ...className
          .toString()
          .split(' ')
          .map(cn => `.${cn}`),
      ];
    }
    return selectors;
  }

  function getStyles(nodeSelectors: string[]) {
    const styles = getStyleSheetTestUtils()
      .getBufferedStyles()
      .join('');
    const ast = css.parse(styles);
    const rules = ast.stylesheet.rules.filter(filter);
    const mediaQueries = getMediaQueries(ast, filter);

    ast.stylesheet.rules = [...rules, ...mediaQueries];

    const ret = css.stringify(ast);

    function filter(rule: css.Rule) {
      if (rule.type === 'rule') {
        return rule.selectors.some(selector => {
          const baseSelector = selector.split(/:| |\./).filter(s => !!s)[0];
          return nodeSelectors.some(
            sel => sel === baseSelector || sel === `.${baseSelector}`,
          );
        });
      }
      return false;
    }

    return ret;
  }

  function getMediaQueries(
    ast: css.Stylesheet,
    filter: (rule: css.Rule) => boolean,
  ) {
    return ast.stylesheet.rules
      .filter(rule => rule.type === 'media' || rule.type === 'supports')
      .reduce((acc, mediaQuery: css.Media | css.Supports) => {
        mediaQuery.rules = mediaQuery.rules.filter(filter);

        if (mediaQuery.rules.length) {
          return [...acc, mediaQuery];
        }

        return acc;
      }, []);
  }

  return {
    test,
    print,
  };
}

// doing this to make it easier for users to mock things
// like switching between development mode and whatnot.
const getAphroditeStyleSheetTestUtils = (): typeof StyleSheetTestUtils =>
  require('aphrodite').StyleSheetTestUtils;

export const aphroditeSerializer = createSerializer(
  getAphroditeStyleSheetTestUtils,
);
