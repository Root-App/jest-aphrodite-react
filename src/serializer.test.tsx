import { css, StyleSheet } from 'aphrodite';
import * as enzyme from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import * as reactTestRenderer from 'react-test-renderer';
import { aphroditeSerializer } from './serializer';

expect.addSnapshotSerializer(aphroditeSerializer);

const Wrapper: React.SFC = props => {
  const styles = StyleSheet.create({
    wrapper: {
      background: 'papayawhip',
      padding: '4em',
    },
  });
  return <section className={css(styles.wrapper)} {...props} />;
};

const Title: React.SFC = props => {
  const styles = StyleSheet.create({
    title: {
      color: 'palevioletred',
      fontSize: '1.5em',
      textAlign: 'center',
    },
  });
  return <h1 className={css(styles.title)} {...props} />;
};

test('react-test-renderer', () => {
  const tree = reactTestRenderer
    .create(
      <Wrapper>
        <Title>
          Hello World, this is my first component styled with aphrodite!
        </Title>
      </Wrapper>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('enzyme', () => {
  const ui = (
    <Wrapper>
      <Title>
        Hello World, this is my first component styled with aphrodite!
      </Title>
    </Wrapper>
  );

  const enzymeMethods = ['shallow', 'mount', 'render'];
  enzymeMethods.forEach(method => {
    const tree = (enzyme as any)[method](ui);
    expect(toJson(tree)).toMatchSnapshot(`enzyme.${method}`);
  });
});

test('works when the root element does not have styles', () => {
  const tree = reactTestRenderer
    .create(
      <div>
        <Wrapper />
      </div>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test(`doesn't fail for nodes without styles`, () => {
  const tree = reactTestRenderer.create(<div />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('joins multiple classes', () => {
  const styles = StyleSheet.create({
    first: { color: 'red' },
    second: { backgroundColor: 'black' },
    third: { color: 'blue' },
  });
  const tree = reactTestRenderer
    .create(<div className={css(styles.first, styles.second, styles.third)} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('supports mediaQueries', () => {
  const styles = StyleSheet.create({
    root: {
      color: 'red',
      '@media(min-width: 200px)': {
        color: 'black',
      },
    },
  });
  const tree = reactTestRenderer
    .create(<div className={css(styles.root)} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('supports pseudo selectors', () => {
  const styles = StyleSheet.create({
    root: {
      color: 'red',
      ':hover': {
        color: 'green',
      },
      '@media(min-width: 200px)': {
        ':hover': {
          color: 'black',
        },
      },
    },
  });
  const tree = reactTestRenderer
    .create(<div className={css(styles.root)} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
