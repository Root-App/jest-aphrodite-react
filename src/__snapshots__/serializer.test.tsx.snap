// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't fail for nodes without styles 1`] = `<div />`;

exports[`enzyme: enzyme.mount 1`] = `
.wrapper {
  background: papayawhip !important;
  padding: 4em !important;
}

.title {
  color: palevioletred !important;
  font-size: 1.5em !important;
  text-align: center !important;
}

<Wrapper>
  <section
    className="wrapper"
  >
    <Title>
      <h1
        className="title"
      >
        Hello World, this is my first component styled with aphrodite!
      </h1>
    </Title>
  </section>
</Wrapper>
`;

exports[`enzyme: enzyme.render 1`] = `
.wrapper {
  background: papayawhip !important;
  padding: 4em !important;
}

.title {
  color: palevioletred !important;
  font-size: 1.5em !important;
  text-align: center !important;
}

<section
  class="wrapper"
>
  <h1
    class="title"
  >
    Hello World, this is my first component styled with aphrodite!
  </h1>
</section>
`;

exports[`enzyme: enzyme.shallow 1`] = `
.wrapper {
  background: papayawhip !important;
  padding: 4em !important;
}

<section
  className="wrapper"
>
  <Title>
    Hello World, this is my first component styled with aphrodite!
  </Title>
</section>
`;

exports[`joins multiple classes 1`] = `
.first-second-third {
  background-color: black !important;
  color: blue !important;
}

<div
  className="first-second-third"
/>
`;

exports[`react-test-renderer 1`] = `
.wrapper {
  background: papayawhip !important;
  padding: 4em !important;
}

.title {
  color: palevioletred !important;
  font-size: 1.5em !important;
  text-align: center !important;
}

<section
  className="wrapper"
>
  <h1
    className="title"
  >
    Hello World, this is my first component styled with aphrodite!
  </h1>
</section>
`;

exports[`supports mediaQueries 1`] = `
.root {
  color: red !important;
}

@media (min-width: 200px) {
  .root {
    color: black !important;
  }
}

<div
  className="root"
/>
`;

exports[`supports pseudo selectors 1`] = `
.root {
  color: red !important;
}

.root:hover {
  color: green !important;
}

@media (min-width: 200px) {
  .root:hover {
    color: black !important;
  }
}

<div
  className="root"
/>
`;

exports[`works when the root element does not have styles 1`] = `
.wrapper {
  background: papayawhip !important;
  padding: 4em !important;
}

<div>
  <section
    className="wrapper"
  />
</div>
`;
